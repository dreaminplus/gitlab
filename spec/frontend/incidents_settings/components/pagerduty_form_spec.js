import { shallowMount } from '@vue/test-utils';
import PagerDutySettingsForm from '~/incidents_settings/components/pagerduty_form.vue';
import { GlAlert } from '@gitlab/ui';

describe('Alert integration settings form', () => {
  let wrapper;
  const resetWebhookUrl = jest.fn();
  const service = { updateSettings: jest.fn().mockResolvedValue(), resetWebhookUrl };

  const findForm = () => wrapper.find({ ref: 'settingsForm' });
  const findWebhookResetBtn = () => wrapper.find('[data-testid="webhook-reset-btn"]');
  const findAlert = () => wrapper.find(GlAlert);

  beforeEach(() => {
    wrapper = shallowMount(PagerDutySettingsForm, {
      provide: {
        service,
        pagerDutySettings: {
          active: true,
          webhookUrl: 'pagerduty.webhook.com',
          webhookUpdateEndpoint: 'webhook/update',
        },
      },
    });
  });

  afterEach(() => {
    if (wrapper) {
      wrapper.destroy();
      wrapper = null;
    }
  });

  it('should match the default snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should call service `updateSettings` on form submit', () => {
    findForm().trigger('submit');
    expect(service.updateSettings).toHaveBeenCalledWith(
      expect.objectContaining({ settingsKey: 'pager_duty_setting_attributes' }),
    );
  });

  describe('Webhook reset', () => {
    it('should show success message and reset webhook url', () => {
      const newWebhookUrl = 'new.webhook.url';
      resetWebhookUrl.mockResolvedValueOnce(newWebhookUrl);
      findWebhookResetBtn().vm.$emit('click');
      return wrapper.vm.$nextTick().then(() => {
        expect(service.resetWebhookUrl).toHaveBeenCalled();
        expect(findAlert().attributes('variant')).toBe('success');
      });
    });

    it('should show error message and NOT reset webhook url', () => {
      resetWebhookUrl.mockRejectedValueOnce();
      findWebhookResetBtn().vm.$emit('click');
      return wrapper.vm.$nextTick().then(() => {
        expect(findAlert().attributes('variant')).toBe('danger');
      });
    });
  });
});
