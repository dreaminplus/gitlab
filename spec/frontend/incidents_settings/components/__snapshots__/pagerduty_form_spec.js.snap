// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert integration settings form Webhook reset should show error message and NOT reset webhook url 1`] = `
<div>
  <!---->
   
  <p>
    Setting up a webhook with PagerDuty will automatically create a GitLab issue for each PagerDuty incident.
  </p>
   
  <form>
    <gl-form-group-stub
      class="col-8 col-md-9 gl-p-0"
    >
      <gl-toggle-stub
        id="active"
        label="Active"
        labelposition="top"
        value="true"
      />
    </gl-form-group-stub>
     
    <gl-form-group-stub
      class="col-8 col-md-9 gl-p-0"
      label="Webhook URL"
      label-class="label-bold"
      label-for="url"
    >
      <div
        class="input-group"
      >
        <gl-form-input-stub
          id="url"
          readonly="true"
          value="pagerduty.webhook.com"
        />
         
        <span
          class="input-group-append"
        >
          <clipboard-button-stub
            cssclass="btn-default"
            text="pagerduty.webhook.com"
            title="Copy"
            tooltipplacement="top"
          />
        </span>
      </div>
       
      <div
        class="gl-text-gray-400 gl-pt-2"
      >
        <gl-sprintf-stub
          message="Create a GitLab issue for each PagerDuty incident by %{docsLink}"
        />
      </div>
       
      <gl-button-stub
        category="tertiary"
        class="gl-mt-3"
        data-testid="webhook-reset-btn"
        icon=""
        size="medium"
        variant="default"
      >
        Reset webhook URL
      </gl-button-stub>
    </gl-form-group-stub>
     
    <gl-button-stub
      category="tertiary"
      class="gl-mt-5 js-no-auto-disable"
      icon=""
      size="medium"
      type="submit"
      variant="success"
    >
      
      Save changes
    
    </gl-button-stub>
  </form>
</div>
`;

exports[`Alert integration settings form should match the default snapshot 1`] = `
<div>
  <!---->
   
  <p>
    Setting up a webhook with PagerDuty will automatically create a GitLab issue for each PagerDuty incident.
  </p>
   
  <form>
    <gl-form-group-stub
      class="col-8 col-md-9 gl-p-0"
    >
      <gl-toggle-stub
        id="active"
        label="Active"
        labelposition="top"
        value="true"
      />
    </gl-form-group-stub>
     
    <gl-form-group-stub
      class="col-8 col-md-9 gl-p-0"
      label="Webhook URL"
      label-class="label-bold"
      label-for="url"
    >
      <div
        class="input-group"
      >
        <gl-form-input-stub
          id="url"
          readonly="true"
          value="pagerduty.webhook.com"
        />
         
        <span
          class="input-group-append"
        >
          <clipboard-button-stub
            cssclass="btn-default"
            text="pagerduty.webhook.com"
            title="Copy"
            tooltipplacement="top"
          />
        </span>
      </div>
       
      <div
        class="gl-text-gray-400 gl-pt-2"
      >
        <gl-sprintf-stub
          message="Create a GitLab issue for each PagerDuty incident by %{docsLink}"
        />
      </div>
       
      <gl-button-stub
        category="tertiary"
        class="gl-mt-3"
        data-testid="webhook-reset-btn"
        icon=""
        size="medium"
        variant="default"
      >
        Reset webhook URL
      </gl-button-stub>
    </gl-form-group-stub>
     
    <gl-button-stub
      category="tertiary"
      class="gl-mt-5 js-no-auto-disable"
      icon=""
      size="medium"
      type="submit"
      variant="success"
    >
      
      Save changes
    
    </gl-button-stub>
  </form>
</div>
`;
